cranelift/fuzzgen/src/function_generator.rs

```diff
@@ -1294,6 +1156,123 @@ where
             .map(|(b, _)| *b)
             .collect();
 
+        // Compute the block CFG
+        //
+        // cranelift-frontend requires us to never generate unreachable blocks
+        // To ensure this property we start by constructing a main "spine" of blocks. So block1 can
+        // always jump to block2, and block2 can always jump to block3, etc...
+        //
+        // That is not a very interesting CFG, so we introduce variations on that, but always
+        // ensuring that the property of pointing to the next block is maintained whatever the
+        // branching mechanism we use.
+        let blocks = self.resources.blocks.clone();
+        self.resources.block_terminators = blocks
+            .iter()
+            .map(|&(block, _)| {
+                // On the last block we always return
+                if block == self.resources.blocks.last().unwrap().0 {
+                    return Ok(BlockTerminator::Return);
+                }
+                let next_block = Block::with_number(block.as_u32() + 1).unwrap();
+
+                // Start with the basic terminators, these can always be inserted, since we
+                // always have at least one forward block
+                let mut terminators: Vec<Box<dyn Fn(&mut Self) -> Result<BlockTerminator>>> = vec![
```

Comment:
The vec of boxed closures here is a little awkward -- is it forced only because we want to use the `Arbitrary::choose()` function below to pick out of a list?

I wonder if it would make sense to instead have something like

```rust
enum BlockTerminatorKind {
  Return,
  Jump,
  Cond,
  BrTable,
  ...
}
```

and then match on that and do generation as required (i.e. have a `match kind { ... }` that evaluates to a `BlockTerminator`)?

Reply:
>The vec of boxed closures here is a little awkward -- is it forced only because we want to use the Arbitrary::choose() function below to pick out of a list?

Yeah, exactly.

> I wonder if it would make sense to instead have something like and then match on that and do generation as required (i.e. have a match kind { ... } that evaluates to a BlockTerminator)?

Sounds Good 👍